satellite/satellitedb/certdb.go

```diff
@@ -19,82 +22,96 @@ type certDB struct {
 	db *dbx.DB
 }
 
-func (certs *certDB) SavePublicKey(ctx context.Context, nodeID storj.NodeID, publicKey crypto.PublicKey) (err error) {
+func (certs *certDB) SavePublicKey(ctx context.Context, nodeID storj.NodeID, pi *identity.PeerIdentity) (err error) {
 	defer mon.Task()(&ctx)(&err)
-	pubbytes, err := pkcrypto.PublicKeyToPKIX(publicKey)
-	if err != nil {
-		return Error.Wrap(err)
-	}
 
-	_, err = certs.db.Get_CertRecord_By_Publickey(ctx, dbx.CertRecord_Publickey(pubbytes))
+	tx, err := certs.db.Begin()
 	if err != nil {
-		if err == sql.ErrNoRows {
-			return certs.tryAddPublicKey(ctx, nodeID, publicKey)
-		}
 		return Error.Wrap(err)
 	}
 
-	// publickey for the nodeID entry already exists, just return
-	return nil
-}
+	defer func() {
+		if err == nil {
+			err = tx.Commit()
+		} else {
+			err = errs.Combine(err, tx.Rollback())
+		}
+	}()
 
-func (certs *certDB) tryAddPublicKey(ctx context.Context, nodeID storj.NodeID, publicKey crypto.PublicKey) (err error) {
-	defer mon.Task()(&ctx)(&err)
-	// no rows err, so create/insert an entry
-	pubbytes, err := pkcrypto.PublicKeyToPKIX(publicKey)
-	if err != nil {
-		return Error.Wrap(err)
+	if pi == nil {
+		return Error.New("Peer Identity cannot be nil")
 	}
+	chain := encodePeerIdentity(pi)
 
-	// TODO: use upsert here instead of create
-	_, err = certs.db.Create_CertRecord(ctx,
-		dbx.CertRecord_Publickey(pubbytes),
-		dbx.CertRecord_Id(nodeID.Bytes()),
-	)
-	// another goroutine might race to create the cert record, let's ignore that error
-	if pgutil.IsConstraintError(err) || sqliteutil.IsConstraintError(err) {
-		return nil
-	} else if err != nil {
+	var id int64
+	query := `SELECT id FROM certRecords WHERE peer_identity = ?;`
```

Comment:
Let's use the serial number to verify whether it needs updating otherwise this will end up really slow.

Reply:
are you saying that there will be only one entry per nodeID? if already a nodeID has an peerIdenitty entry in the table and a new one comes in, will it get updated? or should I make a new entry, because we need that nodeID's previous peer identities as well? what's is the requirement. 

or I am not completely savoy in identities concept, but based on leaf concept and my crude 10000 ft level understanding, if the node(SNO) happened to change its identity, then it just gets added to it previous ones as leaf and keeps the NodeID same. If this is my correct understanding than , there will be only one entry per nodeID and if the peeridentiy column is different , then I just update it. Is my understanding right?